File: code-examples/pymongo/transactions/txn-api-examples.py
from pymongo import MongoClient
from pymongo.errors import PyMongoError

# begin-core
def core_test(client):
    session = client.start_session()
    try:
        session.start_transaction()

        savings_coll = client["bank"]["savings_accounts"]
        savings_coll.update_one(
            {"account_id": "9876"}, 
            {"$inc": {"amount": -100}}, 
            session=session)

        checking_coll = client["bank"]["checking_accounts"]
        checking_coll.update_one(
            {"account_id": "9876"}, 
            {"$inc": {"amount": 100}}, 
            session=session)

        # ... perform other operations

        session.commit_transaction()
        print("Transaction committed.")
    except PyMongoError as error:
        print("An error occurred during the transaction:", error)
        session.abort_transaction()
    finally:
        session.end_session()
# end-core

# begin-conv
def conv_test(client):
    with client.start_session() as session:
        try:
            session.start_transaction()

            savings_coll = client["bank"]["savings_accounts"]
            savings_coll.update_one(
                {"account_id": "9876"}, 
                {"$inc": {"amount": -100}}, 
                session=session)
      
            checking_coll = client["bank"]["checking_accounts"]
            checking_coll.update_one(
                {"account_id": "9876"}, 
                {"$inc": {"amount": 100}}, 
                session=session)

            # ... perform other operations

            session.commit_transaction()
            print("Transaction committed.")

        except PyMongoError as error:
            print("An error occurred during the transaction:", error)
            session.abort_transaction()
# end-conv

def run():
    uri = "<connection string>"
    client = MongoClient(uri)

    try:
        core_test(client)
        conv_test(client)
    finally:
        client.close()
run()

File: code-examples/php/transactions/txn-api-examples.php
<?php

require_once __DIR__ . "/vendor/autoload.php";

use MongoDB\Driver\Manager;
use MongoDB\Driver\Session;
use MongoDB\Driver\Exception\Exception;

// Begin-core
function coreTest($manager) {
    $session = $manager->startSession();

    try {
        $session->startTransaction();

        $coll = $manager->selectCollection('bank', 'savings_accounts');
        $coll->updateOne([ 'account_id' => "9876"], ['$inc' => ['amount' => -100]], ['session' => $session]);

        $coll = $manager->selectCollection('bank', 'checking_accounts');
        $coll->updateOne([ 'account_id' => "9876"], ['$inc' => ['amount' => 100]], ['session' => $session]);

        // ... perform other operations

        $session->commitTransaction();
        echo "Transaction committed.";

    } catch (Exception $e) {
        echo "An error occurred during the transaction:" . $e->getMessage();
        $session->abortTransaction();

    } finally {
        $session->endSession();
    }
}
// End-core

// Begin-conv
function convTest($manager) {
    $session = $manager->startSession();

    $transactionOptions = [
        'readConcern' => new MongoDB\Driver\ReadConcern(MongoDB\Driver\ReadConcern::MAJORITY),
        'writeConcern' => new MongoDB\Driver\WriteConcern(MongoDB\Driver\WriteConcern::MAJORITY),
    ];

    $callback = function(Session $session) use ($manager) {
        $coll = $manager->selectCollection('bank', 'savings_accounts');
        $coll->updateOne([ 'account_id' => "9876"], ['$inc' => ['amount' => -100]], ['session' => $session]);

        $coll = $manager->selectCollection('bank', 'checking_accounts');
        $coll->updateOne([ 'account_id' => "9876"], ['$inc' => ['amount' => 100]], ['session' => $session]);

        // ... perform other operations      

        return "Transaction committed.";
    };

    echo $manager->executeTransaction($session, $callback, $transactionOptions);

    $session->endSession();
}
// End-conv

function run() {
    $uri =
        "<connection string>";
    $manager = new MongoDB\Driver\Manager($uri);

    try {
        coreTest($manager);
        convTest($manager);

    } finally {
        $manager->close();
    }
}
run();

?>
